common/src/main/java/io/homeassistant/companion/android/common/data/prefs/WearPrefsRepositoryImpl.kt

```
@@ -52,15 +54,64 @@ class WearPrefsRepositoryImpl @Inject constructor(
         }
     }
 
-    override suspend fun getTileShortcuts(): List<String> {
-        val jsonArray = JSONArray(localStorage.getString(PREF_TILE_SHORTCUTS) ?: "[]")
-        return List(jsonArray.length()) {
-            jsonArray.getString(it)
+    override suspend fun getTileShortcutsAndSaveTileId(tileId: Int): List<String> {
+        val tileIdToShortcutsMap = getAllTileShortcuts()
+        return if (null in tileIdToShortcutsMap && tileId !in tileIdToShortcutsMap) {
+            // if there are shortcuts with an unknown (null) tileId key from a previous installation,
+            // and the tileId parameter is not already present in the map, associate it with those shortcuts
+            val entities = removeTileShortcuts(null)!!
+            setTileShortcuts(tileId, entities)
+            entities
+        } else {
+            val entities = tileIdToShortcutsMap[tileId]
+            if (entities == null) {
+                setTileShortcuts(tileId, emptyList())
+            }
+            entities ?: emptyList()
         }
     }
 
-    override suspend fun setTileShortcuts(entities: List<String>) {
-        localStorage.putString(PREF_TILE_SHORTCUTS, JSONArray(entities).toString())
+    override suspend fun getAllTileShortcuts(): Map<Int?, List<String>> {
+        return localStorage.getString(PREF_TILE_SHORTCUTS)?.let { jsonStr ->
+            runCatching {
```

Why try the 'old' way first, instead of the other way around like you did originally? This means we will always encounter an exception first, when the preference is updated.

---

Good point! Originally my concern was that if the backward compatibility part is in the `recover` block, and we don't do anything with the `Exception`, then random other `Exception`s happening in the `runCatching` block would be silently swallowed.

I refactored it in 2d922ae1.